### DISCUSSION TITLE
Coerce makes the schema 'nullable' by default

### DISCUSSION THREAD

**Discussion by ernesto-elder** [colinhacks/zod#2001] [Q&A] [answered]
## Description
[Description of the issue, including context and why it's a problem]
When using coercion for any value, the resulting schema is nullable by default. This is a problem because I couldn't find a way to revert the 'nullable' status.

## Steps to Reproduce
1. Create a schema using the `coerce` attribute, like `const stringSchema = z.coerce.string()`

## Expected Behavior
`stringSchema.isNullable() === false`

## Actual Behavior
`stringSchema.isNullable() === true`

## Additional Information
Here is a [demo](https://codesandbox.io/s/zod-demo-forked-ymootx?file=/src/index.ts) to demonstrate the issue

**Comment by JacobWeisenburger** [comment]
This seems to be working as expected for me.
```ts
const stringSchema = z.coerce.string()
console.log( stringSchema.safeParse( null ) )
// { success: true, data: 'null' }
```

Can you explain why you don't want the `stringSchema` to accept `null`, but yet allow anything else?

**Comment by ernesto-elder** [comment]
> This seems to be working as expected for me.
> 
> ```ts
> const stringSchema = z.coerce.string()
> console.log( stringSchema.safeParse( null ) )
> // { success: true, data: 'null' }
> ```
> 
> Can you explain why you don't want the `stringSchema` to accept `null`, but yet allow anything else?

Ah... I see it now, of course. Coerce will take ANYTHING and try to transform it into the required value, which makes sense.  

The problem I was having is that I'm creating a form and the date field can be either nullable or not, and based on that I'm doing one thing or the other. To know if the field is nullable, I'm taking the schema and calling `theDateSchema.isNullable()` and, because I'm using coerce on the date field, this always returns `true`.

I'm getting around this problem by doing the validation manually, like so:

```js
 const dateSchema = z.string().transform((date, ctx) => {
  // validate the date here and register a validation issue with 'addIssue'
})
```
In this example, `dateSchema.isNullable() === false` by default, and, if I want it to be nullable, I'd just call the `nullable()` method when creating the schema.

If there's a better way of doing this, would you please share it? 😃 

Thanks

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
const dateSchema = z.date().or( z.string() ).pipe( z.coerce.date() )
console.log( dateSchema.isNullable() ) // false
console.log( dateSchema.safeParse( '2021-01-01' ).success ) // true
console.log( dateSchema.safeParse( new Date ).success ) // true
console.log( dateSchema.safeParse( null ).success ) // false

const nullableDateSchema = z.date().or( z.string() ).pipe( z.coerce.date() ).nullable()
console.log( nullableDateSchema.isNullable() ) // true
console.log( nullableDateSchema.safeParse( '2021-01-01' ).success ) // true
console.log( nullableDateSchema.safeParse( new Date ).success ) // true
console.log( nullableDateSchema.safeParse( null ).success ) // true
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by ernesto-elder** [reply]
Yes! That's exactly what I'm looking for, thanks! 🎉  I didn't know about the `pipe` method though, I can't find it in the docs either. I can see on the releases page that was recently introduced in version 3.20. Pretty cool.

**Reply by JacobWeisenburger** [reply]
Thanks so much for the sponsorship!
